versions/3.0.4.md

```diff
@@ -3026,6 +3026,8 @@ When request bodies or response payloads may be one of a number of different sch
 This hint can be used to aid in serialization, deserialization, and validation.
 The Discriminator Object does this by implicitly or explicitly associating the possible values of a named property with alternative schemas.
 
+Note that `discriminator` MUST NOT change the validation outcome of the schema.
```

Comment:
This is surprising: if `discriminator` is a hint for picking the "right" schema out of an `anyOf` array, the result may differ if a validator uses the hint and successfully validates (only) against the hinted alternative, or if it sequentially tries all alternatives and stops on the first success, which may not be the one `discriminator` hints at.

Do we indirectly say here that validators MUST ignore `discriminator`?

Reply:
Does JSON Schema proscribe an order in which the schemas of an `anyOf` are checked? I did not think so but I'm certainly no expert here.

Reply:
[Draft 04](https://json-schema.org/draft-04/draft-fge-json-schema-validation-00#rfc.section.5.5.4) only says

> An instance validates successfully against this keyword if it validates successfully against at least one schema defined by this keyword's value.

So my implementation would loop over the schemas in array order and stop at the first success. Why waste energy on the remaining schemas.

[2020-12](https://json-schema.org/draft/2020-12/json-schema-core#section-10.2.1.2) requires that

> all subschemas MUST be examined

As this PR is for 3.0.4 I would go with my interpretation of Draft-04.

Reply:
Okay, so I don't think there is a requirement for validators to ignore the `discriminator`. If JSON Schema allows validators to choose the order of validation, then a validator is free to choose to start with the schema associated with the discriminator, or not.

Reply:
@ralfhandl 

> Do we indirectly say here that validators MUST ignore `discriminator`?

No, if you know what sort of static analysis of the entire schema that you need to do to calculate whether an `anyOf` branch is safe to skip _and_ you don't support annotations (which is sad because annotations are the easiest way to extend JSON Schema), you can technically use it to short-circuit.  I think @karenetheridge 's implementation does work out whether one of the annotation-consuming keywords like `unevaluatedProperties` is involved, for example, although I think she just uses it to decide whether it's necessary to collect property annotations for certain subschemas.

Reply:
@ralfhandl 

> That implementaiton is incorrect for draft 2020-12. It's fine up through draft-07, but not after that.

I'm having a slow morning and have been forward-porting things to 3.1.1, so I was in a 3.1 mindset.  But this is on 3.0.4 🤦 

So in 3.0.4 it's entirely reasonable to short-circuit `anyOf` because draft-04 did not define how to collect annotations (much less have keywords relying on that sort of thing).

But that _still_ doesn't change the validation outcome.  I was very specific in my word choice of "validation outcome" as there are many other things about schema evaluation than just the validation outcome.